### ISSUE TITLE
[🐛 Bug]: Memory leak on any jre and huge amount of threads on jre 17 with jdk http client .

### ISSUE THREAD

**Issue by pogrebnyaksa** [SeleniumHQ/selenium#12921]
### What happened?

My application spawn multiple threads creating remoteWebriver connecting to selenium standalone chrome. Connection is established using jdk http client. I experence 2 issues:

1.  On jre 17 I can see huge amount threads not stopping after minutes or hours passed. (insert image from repro with threads graph for jre 17) Problem with threads is not observed on jre 20.  (insert image from repro with threads graph for jre 17). And currently I can not use jre 20 (not LTS) or 21 (not production ready, just GA)

2. Memory leak. If you run my repro for 5 mins you can see that memory usage is increased. After some time my repro stops connecting to selenium standalone and according to docker stats you can see that memory is not freed up, even after performing full gc using peform GC in jconsole. Issue is observed on jre 17,20.

This issue relates to #11270

### How can we reproduce the issue?
Please see [repo:](https://github.com/pogrebnyaksa/selenium-repo2)

### Relevant log output

```shell
no log output required
```

### Operating System

Debian 9, Ubuntu 22.04.3

### Selenium version

Java 4.13.0

### What are the browser(s) and version(s) where you see this issue?

Google Chrome 117.0.5938.132

### What are the browser driver(s) and version(s) where you see this issue?

 Drirver in selenium standalone image 4.13 117.0.5938.92

### Are you using Selenium Grid?

no, using selenium standalone chrome docker 4.13 image

**Comment by diemol** [comment]
Can you reproduce this only using Selenium? I see several different dependencies on your repository.

**Comment by pogrebnyaksa** [comment]
> Can you reproduce this only using Selenium? I see several different dependencies on your repository.

The same result. I've added new branch, please take a look there [branch-java-only](https://github.com/pogrebnyaksa/selenium-repo2/tree/java-only)

**Comment by pujagani** [comment]
Thank you for sharing the details. I have used the same code though I skipped adding the Chrome arguments. https://github.com/pogrebnyaksa/selenium-repo2/blob/java-only/src/main/java/com/example/selrepro/SelReproApplication.java

I am using the Selenium Grid 4.14.0 version which defaults to Java 11 HttpClient and running the jar locally. I am unable to see a memory leak pattern. I ran the code for about 30 minutes duration. I am on openjdk version "17.0.8.1".
Sharing the snapshot below: 

![Screenshot 2023-10-13 at 4 31 58 PM](https://github.com/SeleniumHQ/selenium/assets/10705590/73668c2f-6a0d-40dd-95a1-3bc0aff54bbd)

**Comment by pogrebnyaksa** [comment]
With heap I don't observe issues either, but if you check total RAM consumption, you can see that app consumes all ram
check docker stats
NAME              CPU %     MEM USAGE / LIMIT   MEM %        
 sel-repro-app-1   7.81%     499.6MiB / 500MiB   99.93%    
 but, I have tried to free native heap using jmx method systemTrimNativeHeap , it frees around 60mb 
 NAME              CPU %     MEM USAGE / LIMIT   MEM %     
sel-repro-app-1   0.76%     438.9MiB / 500MiB   87.78% 
I have tried with such version of jre:
openjdk version "17.0.8.1" 2023-08-24
OpenJDK Runtime Environment Temurin-17.0.8.1+1 (build 17.0.8.1+1)
OpenJDK 64-Bit Server VM Temurin-17.0.8.1+1 (build 17.0.8.1+1, mixed mode, sharing)
May be there is native memory leak? 
also in 17.0.9 https://bugs.openjdk.org/browse/JDK-8293114 we will receive autotrim of native heap. 

Regarding issue with big amount of threads - on my machine the number of them doesn't lower after spawning of selenium instances stopped. I have waited for about 10 mins after. 
I have tried running on selenium 4.13, 4.14, received same results. Selenium standalone chrome 4.14 was not released  yet.

**Comment by pujagani** [comment]
Thank you for sharing the details and trying the new versions. I have checked the RAM used this time. After running the code snippet above for two sessions (30 minutes, gap of a few minutes, 30 minutes), I see the RAM used increases. However, the increase has not led to OOM or grown exponentially. I tried running more tests in parallel, essentially maxing out Grid's capacity of running parallel sessions and still the RAM used has not grown exponentially. 

The memory keeps increasing gradually as we run more sessions and ideally, that shouldn't happen so I suspect a potential memory leak. 

I am going to run a similar experiment with newer versions of Java and report my findings here.

The above tests are run using  Grid version 4.14, 4.14.1 , Openjdk version "17.0.8.1".

I am sharing my findings below for Selenium 4.14. Selenium 4.14.1 also has similar numbers. 
PID is 16178. The memory used at the start when no sessions are running is  ~120 MB.
After 30 minutes of running sessions, it is ~274 MB. This increases further if the code snippet is run for another 30 minutes.

![Screenshot 2023-10-18 at 10 55 52 AM](https://github.com/SeleniumHQ/selenium/assets/10705590/0618daad-c1cc-4bc2-a5cb-847df58083f8)

![Screenshot 2023-10-18 at 11 26 22 AM](https://github.com/SeleniumHQ/selenium/assets/10705590/b0276b40-00ae-4547-83e0-6cfaa65db58b)

However, I have not seen the problem regarding the threads mentioned in either of the cases. 

Meanwhile, are you aware of a specific Selenium version when you started to see the memory leak?

**Comment by pujagani** [comment]
I have repeated the same steps but with Java 21.0.1. I am seeing a similar pattern as mentioned earlier.

**Comment by pujagani** [comment]
The concern here is RAM used gradually increases as you run more tests. I will try digging into this.

**Comment by pujagani** [comment]
I ran the same code for 6 hours yesterday evening, however, I did not see the RAM increasing, it was constant after a point. Triaging this would be convenient if I could reproduce this constantly or rather quickly. I have taken the native memory baseline snapshot and a snapshopt of the same after running the code for 30 minutes. I tried analyzing the difference in the too but will need to revisit to find a glaring increase.
